runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/BeamWorkerStatusGrpcService.java

```diff
@@ -0,0 +1,190 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamWorkerStatusGrpcService.class);
+  private static final long DEFAULT_CLIENT_CONNECTION_WAIT_TIME_SECONDS = 5;
+  private final HeaderAccessor headerAccessor;
+  private final Map<String, CompletableFuture<WorkerStatusClient>> connectedClient =
+      new ConcurrentHashMap<>();
+
+  private BeamWorkerStatusGrpcService(
+      ApiServiceDescriptor apiServiceDescriptor, HeaderAccessor headerAccessor) {
+    this.headerAccessor = headerAccessor;
+    LOG.info("Launched Beam Fn Status service at {}", apiServiceDescriptor);
+  }
+
+  /**
+   * Create new instance of {@link BeamWorkerStatusGrpcService}.
+   *
+   * @param apiServiceDescriptor describes the configuration for the endpoint the server will
+   *     expose.
+   * @param headerAccessor headerAccessor gRPC header accessor used to obtain SDK harness worker id.
+   * @return {@link BeamWorkerStatusGrpcService}
+   */
+  public static BeamWorkerStatusGrpcService create(
+      ApiServiceDescriptor apiServiceDescriptor, HeaderAccessor headerAccessor) {
+    return new BeamWorkerStatusGrpcService(apiServiceDescriptor, headerAccessor);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (CompletableFuture<WorkerStatusClient> clientFuture : connectedClient.values()) {
+      if (clientFuture.isDone()) {
+        clientFuture.get().close();
+      }
+    }
+    connectedClient.clear();
+  }
+
+  @Override
+  public StreamObserver<WorkerStatusResponse> workerStatus(
+      StreamObserver<WorkerStatusRequest> requestObserver) {
+    String workerId = headerAccessor.getSdkWorkerId();
+    LOG.info("Beam Fn Status client connected with id {}", workerId);
+
+    WorkerStatusClient fnApiStatusClient =
+        WorkerStatusClient.forRequestObserver(workerId, requestObserver);
+    fnApiStatusClient.setDeregisterCallback(this.connectedClient::remove);
+    if (connectedClient.containsKey(workerId) && connectedClient.get(workerId).isDone()) {
+      LOG.info(
+          "SDK Worker {} was connected to status server previously, disconnecting the old client",
+          workerId);
+      try {
+        WorkerStatusClient oldClient = connectedClient.get(workerId).get();
+        oldClient.close();
+      } catch (IOException | InterruptedException | ExecutionException e) {
+        LOG.warn("Error closing worker status client", e);
+      }
+    }
+    connectedClient
+        .computeIfAbsent(workerId, k -> new CompletableFuture<>())
+        .complete(fnApiStatusClient);
+    return fnApiStatusClient.getResponseObserver();
+  }
+
+  /**
+   * Get the latest SDK worker status from the client's corresponding SDK Harness.
+   *
+   * @param workerId worker id of the SDK harness.
+   * @return {@link CompletableFuture} of WorkerStatusResponse from SDK harness.
+   */
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus(String workerId) {
```

Comment:
I'm not sure this method will provide much value since all implementations will attempt to get all statuses.

Reply:
I've moved this to a private method and added another getSingleWorkerStatus() function. The idea is that we can provide user a parameterized query which can return single SDK harness status.